src/core/common/string.hpp

```diff
@@ -0,0 +1,215 @@
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *  This file defines OpenThread String class.
+ */
+
+#ifndef STRING_HPP_
+#define STRING_HPP_
+
+#include "openthread-core-config.h"
+
+#include <stdarg.h>
+#include <stdio.h>
+#include "utils/wrap_string.h"
+
+#include <openthread/types.h>
+
+#include "common/code_utils.hpp"
+
+namespace ot {
+
+/**
+ * @addtogroup core-string
+ *
+ * @brief
+ *   This module includes definitions for OpenThread String class.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This class defines the base class for `String`.
+ *
+ */
+class StringBase
+{
+protected:
+    /**
+     * This method appends `printf()` style formatted data to a given character string buffer.
+     *
+     * @param[in]    aBuffer  A pointer to buffer containing the string.
+     * @param[in]    aSize    The size of the buffer (in bytes).
+     * @param[inout] aLength  A reference to variable containing current length of string. On exit length is updated.
+     * @param[in]    aFormat  A pointer to the format string.
+     * @param[in]    aArgs    Arguments for the format specification.
+     *
+     * @retval OT_ERROR_NONE           Updated the string successfully.
+     * @retval OT_ERROR_NO_BUFS        String could not fit in the storage.
+     * @retval OT_ERROR_INVALID_ARGS   Arguments do not match the format string.
+     */
+    static otError Write(char *aBuffer, uint16_t aSize, uint16_t &aLength, const char *aFormat, va_list aArgs);
+};
+
+/**
+ * This class defines a fixed-size string.
+ *
+ */
+template <uint16_t SIZE> class String : private StringBase
```

Comment:
What is the code size / flash impact of this usage of templates?

Reply:
The main functionality is implemented by the common base class `StringBase`. The template variations use the common base class method.

In general, this is used for logging purposes. With logging level as default as CRIT, there is no change in code size (We don't pay for what we don't use). With `INFO` log level enabled there is small difference ~250 bytes.

````
   text	   data	    bss	    dec	    hex	filename
 216123	   4588	  22264	 242975	  3b51f	output/cc2538/bin/ot-ncp-ftd  (Log level CRIT - master)
 216123	   4588	  22264	 242975	  3b51f	output/cc2538/bin/ot-ncp-ftd  (Log level CRIT - pr/string2 branch) 
 236204	   4588	  22264	 263056	  40390	output/cc2538/bin/ot-ncp-ftd  (Log level INFO - master)
 236432	   4588	  22264	 263284	  40474	output/cc2538/bin/ot-ncp-ftd  (Log level INFO - pr/string2 branch)
````